Feature/2 diary list

pozafly opened the pull request:
## 주제

- 다이어리 카드 목록 페이지 구현

## 작업 내용

1. 공통 컴포넌트 수정 및 제작
	- getLayout 패턴 적용
	- Header 제작
	- Dropdown 제작
		- Compound Component 패턴 적용
		- input Validation 적용
		- useClickOutSide hooks 제작
2. diary card list 페이지 제작
	- react query로 조회, 생성, 삭제 로직 적용
	- 로딩 스피너 타이밍 적용
3. 프로젝트 폴더 구조를 features 단위로 재작성

## ⭐️ 멘토님 코멘트 필요

- Dropdown 컴포넌트에 패턴 적용이 잘 되었는지?
- 프로젝트 폴더 구조가 괜찮은지?

## 캡쳐

- 카드 리스트 생성
![화면 기록 2023-07-03 오전 11 52 50](https://github.com/f-lab-edu/jiary/assets/59427983/07a6629e-7c9e-407e-b06f-f1913f600d88)

- 카드 삭제
![화면 기록 2023-07-03 오전 11 59 21](https://github.com/f-lab-edu/jiary/assets/59427983/a9fb33a4-345b-4236-b09c-513b97d723a6)

pozafly committed:
feat: interceptor에 요청을 보낼 때마다 header에 accessToken을 넣어서 보내도록 수정

pozafly committed:
chore: add next config external image path

pozafly committed:
feat: add common header, profile components

pozafly committed:
fix: auth slice selector isLoggedInSelector boolean 값으로 변경

pozafly committed:
refactor: global css 변경

pozafly committed:
fix: PageLoadingSpinner extendsion tsx -> ts

pozafly committed:
feat: add page common Layout apply

pozafly committed:
refactor: page layout css 수정

pozafly committed:
refactor: page layout css 수정

pozafly committed:
refactor: page header css 수정

pozafly committed:
feat: add diary page apis

pozafly committed:
feat: diary page 퍼블리싱

pozafly committed:
feat: add common dropdown apply compound components pattern

pozafly committed:
feat: add constant color system

pozafly committed:
feat: 다이어리 목록 페이지 헤더 추가 및 드롭다운 적용

pozafly committed:
fix: 드롭다운 컴포넌트 외부 클릭시 닫히면 input value 초기화 되도록 수정

pozafly committed:
refactor: 드롭다운 컴포넌트 외부에서 input 값 주입하도록 변경

pozafly committed:
feat: 다이어리 생성 로직 및 로직 연결

pozafly committed:
fix: 다이어리 리스트 example data 제거

pozafly committed:
chore: tsconfig.json target을 es5 -> es6. iterator를 사용하기 위해

pozafly committed:
fix: 다이어리 카드 제목 jiary- 제거하여 보여주도록 수정

pozafly committed:
feat: 드롭다운 input에 enterkey 이벤트 바인딩

pozafly committed:
feat: 다이어리 리스트 페이지 로딩 스피너 추가

pozafly committed:
fix: page loading spinner 디자인 변경

pozafly committed:
fix: auth page css and add auth page router

pozafly committed:
fix: page loading spinner css 수정

pozafly committed:
chore: add vanilla-extract/dynamic

pozafly committed:
refactor: 드롭다운 컴포넌트 props 추가 및 vanilla-extract/dynamic 활용

pozafly committed:
chore: googleapis api/auth.ts에 scope(drive) 추가

pozafly committed:
feat: 다이어리 삭제 로직 api 생성

pozafly committed:
refactor: 다이어리 리스트 카드에 삭제 드롭다운 리팩토링 및 css 수정

pozafly committed:
feat: 다이어리 리스트 페이지 삭제 로직 추가 및 로딩 스피너 연동

pozafly committed:
fix: response 204 is not response message

pozafly committed:
fix: loading spinner 전역적으로 동작하도록 layout으로 옮기고, redux로 제어하도록 수정

pozafly committed:
feat: mutation 이후 카드 리스트 화면 즉시 반영되도록 수정

pozafly committed:
rename: diary list 페이지의 컴포넌트를 features/diary/diaryList로 이동

pozafly committed:
rename: diary list 페이지의 컴포넌트를 features/diaryList로 이동

f-lab-james commented on src/features/diaryList/components/DiaryCard.tsx:
```diff
@@ -0,0 +1,50 @@
+      target.className.includes('deleteButton')
+    ) {
+      e.preventDefault();
+    }
```
이렇게 컴포넌트랑 style을 한 번 더 묶어주면 좋을 것 같아요!

```
ㄴ components
  ㄴ DiplayCard
    ㄴ index.ts
    ㄴ styles.css.ts 
  ㄴ DispalyList
  ㄴ DisplayHeader
```

f-lab-james commented on src/pages/auth/index.css.ts:
```diff
@@ -1,8 +1,8 @@
+export const container = style({
   position: 'relative',
-  height: '100vh',
+  height: 'calc(100vh - 56px)',
```
pages 밑에 있으면 파일 시스템 기반 라우터여서 /auth/index.css로 걸릴거에요 (정확하진 않지만 어쨋든 pages 밑에 있다는건 라우터랑 연결된다는 뜻이라서 스타일 관련된건 다른쪽에 두셔야합니다.)

f-lab-james commented on src/pages/auth/index.tsx:
```diff
@@ -1,27 +1,15 @@
 import * as style from '@/pages/auth/index.css.ts';
```
css를 import 하기 위해 2가지 방법으로 위치시킬 수 있습니다.

1. features/styles 폴더로 관리
2. features/pages를 만들고 pages에서는 아래와 같이 사용


Path: pages/auth/index.tsx
```ts
export { default } from '@/feature/auth/pages/index'

async function getServerSideProps(context) {
  // ...
}
```

이렇게 두면 페이지 컴포넌트랑 스타일 컴포넌트를 같은 폴더에 둘 수 있습니다.
- feature/auth/pages/auth/index.ts
- feature/auth/pages/auth/index.css.ts

f-lab-james commented on src/features/auth/hooks/useAuth.ts:
```diff
@@ -59,6 +61,7 @@ export const useAuth = () => {
 
     popupWindow?.close();
     window.removeEventListener('message', messageCallback, false);
+    router.push('/diary');
```
위에 URL 대신 URLSearchParams로 해결할 수 있는지도 확인해볼 것!

ref) https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams

f-lab-james commented on src/constant/colors.ts:
```diff
@@ -0,0 +1,17 @@
+  RED_LIGHT_COLOR: 'hsl(11, 76%, 59%)',
+
+  ORANGE_COLOR: 'hsl(32, 100%, 50%)',
+  ORANGE_LIGHT_COLOR: 'hsl(32, 100%, 55%)',
```
https://vanilla-extract.style/documentation/api/create-var/

f-lab-james commented on src/features/common/hooks/useClickOutSide.ts:
```diff
@@ -0,0 +1,27 @@
+import { useCallback, useEffect, useRef } from 'react';
+
+export const useClickOutSide = (callback: () => void) => {
+  const containerRef = useRef(null);
```
containerRef는 외부에서 주입받는 방식으로 수정하는게 더 좋아보이는데 어떨까요?

f-lab-james commented on src/features/diaryList/components/DiaryCardDropdown.tsx:
```diff
@@ -0,0 +1,40 @@
+      target={{ targetElement, targetRef }}
+      control={{ isDropdownOpen, setIsDropdownOpen }}
+    >
+      <button onClick={handleDeleteButton} className={style.deleteButton}>
```
컴파운드 컴포넌트 패턴을 사용하는 목적 중 하나가 Native HTML Element과 유사한 경험을 얻고자 하는 것이기 때문에 props를 이용해서 컴포넌트를 넘기는 것은 최대한 지양해주세요!

```tsx
<Dropdown.Root>
  <Dropdown.Trigger onClick={() => setIsDropdownOpen(!isDropdownOpen)} />
</Dropdown.Root>
```

f-lab-james commented on src/features/diaryList/components/DiaryList.tsx:
```diff
@@ -0,0 +1,46 @@
+  const dispatch = useDispatch();
+
+  const isFetching = useIsFetching({ queryKey: [DIARY_KEY] });
+  const isMuatating = useIsMutating({ mutationKey: [DIARY_KEY] });
```
loading spinner가 전역에서 사용되는 UI이기 때문에 위치가 여기 있을 필요가 없고 _app.tsx에 두어도 될 것 같아요!

f-lab-james commented on src/features/diaryList/components/DiaryListHeader.tsx:
```diff
@@ -0,0 +1,62 @@
+  const handleOpenDropdown = () => {
+    setIsDropdownOpen(isDropdownOpen => !isDropdownOpen);
+    return;
+  };
```
```suggestion
  const handleOpenDropdown = () => setIsDropdownOpen(isDropdownOpen => !isDropdownOpen)
```

f-lab-james commented on src/features/diaryList/components/DiaryListHeader.tsx:
```diff
@@ -0,0 +1,62 @@
+  };
+
+  const handleCreateDoc = () => {
+    setIsDropdownOpen(false);
```
submit 할 때 Dropdown이 닫아주는게 외부에서 컨트롤 할 필요가 없어 보입니다 ex) select

f-lab-james commented on src/pages/index.tsx:
```diff
@@ -1,39 +1,19 @@
-      </main>
+
+      <div className={style.Main}>Home</div>
+      {isPageLoading ? <PageLoadingSpinner /> : null}
```
```tsx
{ isPageLoading && <PageLoadingSpinner /> }
```

pozafly committed:
chore: tsconfig target es6 -> es5. iteration 관련 잘못된 설정을 되돌림

pozafly committed:
rename: src/constant -> src/constants

pozafly committed:
rename: 모든 features 폴더의 apis 구조 변경

- queries.ts, mutations.ts 파일에 들어있던 axios 관련 코드를 index.ts로 옮김
- queries, mutations 폴더를 따로 만들고, react-query 훅을 각각의 파일로 분리함

pozafly committed:
feat: jiary instance를 생성 후 이어줌

pozafly committed:
rename: components 구조에 폴더를 하나 두어, style과 component들을 구분할 수 있도록 변경

pozafly committed:
rename: pages 아래에 있던 라우팅 컴포넌트를, 데이터 조회만 하도록 변경 및 features 하위에 pages 폴더 생성

pozafly committed:
refactor: useClickOutSide 훅을 ref를 받도록 변경

pozafly committed:
refactor: global loading spinner 적용 및 redux에서 loading 상태 제거

pozafly committed:
refactor: 화살표 함수에서 중괄호 및 return 문 제거

pozafly committed:
chore: add jwt-decode libarary

pozafly committed:
fix: api/auth 재 로그인 시 refreshToken 없어지는 문제를 userEmail을 생성해 해결

pozafly committed:
refactor: Dropdown Compound Components 패턴 리팩토링

- 기존 nativeElement 처럼 사용하지 않았던 부분, 모두 nativeElement처럼 사용하도록
- 외부에서 input값과 isShow 상태 값 다루던 것을 내부 처리
- 여닫이 component인 Trigger 컴포넌트 추가

pozafly committed:
refactor: Dropdown 컴포넌트 주요 로직 hooks로 분리

pozafly committed:
fix: Dropdown의 Trigger 컴포넌트 렌더링 시점에 setState 로직 경고 수정

pozafly committed:
refactor: Dropdown 컴포넌트 Input Validation을 validation props 하나로 통일

pozafly committed:
refactor: Dropdown 컴포넌트의 trigger component를 callback ref로 변경

pozafly committed:
refactor: Dropdown 컴포넌트의 hooks 제거 및 context 격리, eslint 빈 함수 rule 수정
